### PR TITLE
MB-47265: Incorrect accounting for JSON arrays while indexing

### PR DESCRIPTION
**Pull request by abhinavdangeti**
+ For JSON array fields, multiple tokens can be added for a single
  field.
+ In the current code, we seem to just overwrite the index of the
  token frequencies - with which only the last written token is
  search-able.
+ Fix comes with a unit test.

### CODE REVIEW

File: doc.go [hunk 1]
```diff
@@ -61,8 +59,16 @@ func (d *Document) newField(field index.Field) {
...
-	d.fieldTokenFreqs = append(d.fieldTokenFreqs, af)
-	d.fieldLens = append(d.fieldLens, field.AnalyzedLength())
+	fieldIdx, exists := d.fieldIndexes[field.Name()]
+	if !exists {
+		d.fieldIndexes[field.Name()] = len(d.fieldTokenFreqs)
+		d.fieldNames = append(d.fieldNames, field.Name())
+		d.fieldTokenFreqs = append(d.fieldTokenFreqs, af)
+		d.fieldLens = append(d.fieldLens, field.AnalyzedLength())
+	} else {
+		d.fieldTokenFreqs[fieldIdx].MergeAll(field.Name(), af)
```
Comment (mschoch) [inline doc.go:69]: This is the only part I'm not sure about.  The issue is that MergeAll does something special with the remoteField argument that really is only intended for building up the composite field data.  My concern is that this subtlely breaks something, although I cannot explain what...

**Review by mschoch — COMMENTED** [review]
Looks fine, except for the MergeAll() call which might be right, but I'm not sure..

File: doc.go (hunk 1 above)
Comment (abhinavdangeti) [reply doc.go:69]: Ya fair point.

Let me create a simple version of merge token frequencies to here.

**Comment by mschoch** [comment]
Sorry pushing all the wrong buttons here, was not trying to resolve, and then tried to request re-review of myself and did it for @sreekanth-cb instead.

**Comment by mschoch** [comment]
So, instead of copying that function, I was thinking we could simply pass empty string, or the field name, whichever is correct, and continue using the existing function.  Ideally, also improving the documentation for this method, with regard to that parameters correct use.

**Comment by mschoch** [comment]
I think the thing that breaks is if you search on the _all field for these matches, does highlighting work?  I think in the past we've accidentally broken this once or twice, and it has that weird symptom, because we find the matches, but the returned location data doesn't realize it was from the field or something like that...

**Comment by abhinavdangeti** [comment]
Hmm, just looked up the bleve code on this. Seems like using the field name is the right thing to do.

**Comment by mschoch** [comment]
> Hmm, just looked up the bleve code on this. Seems like using the field name is the right thing to do.

Can you share what you reviewed to come to this conclusion?

**Comment by abhinavdangeti** [comment]
Upsidedown's analysis is pretty identical - https://github.com/blevesearch/bleve/blob/master/index/upsidedown/analysis.go#L57...L82

**Review by sreekanth-cb — APPROVED** [review]
Looks OK to me.

**Pull request merged**